### PR TITLE
[codex] Move SCF converger ML assets into OpenQP

### PR DESCRIPTION
**Pull request by karmachoi**
## Summary

- Move the `karmachoi/openqp-scf-converger-ml` companion project into `tools/scf-converger-ml/` inside OpenQP, including the training database, contribution collector, geometries, harness scripts, training script, distilled model, and metrics.
- Add an OpenQP-relative `train-scf-selector` GitHub Actions workflow so data changes retrain the model and sync the runtime copy in `pyoqp/oqp/library/scf_selector_model.py`.
- Normalize ML selector labels for DIIS subtypes (`cdiis`, `ediis`, `adiis`, `vdiis`) so `converger_type=ml` sets native `diis` plus the selected DIIS subtype instead of silently losing subtype predictions.

## Validation

- `tools/scf-converger-ml/train/run.sh` using a temporary local Python environment with `scikit-learn` and `numpy`: generated the selector, reported LOO-by-system accuracy `0.62`, and synced the runtime model.
- `/tmp/openqp-scf-move.96Md0v/venv/bin/python -m pytest tests/test_single_point_scf_fallback.py tests/test_scf_selector_training_assets.py` -> `13 passed`.
- `/tmp/openqp-scf-move.96Md0v/venv/bin/python -m py_compile tools/scf-converger-ml/contribute/collect.py tools/scf-converger-ml/harness/build_database.py tools/scf-converger-ml/harness/build_selector.py tools/scf-converger-ml/harness/ensemble_scf_proto.py tools/scf-converger-ml/harness/reparse_fock.py tools/scf-converger-ml/harness/systems.py tools/scf-converger-ml/train/train_selector.py pyoqp/oqp/library/scf_selector_model.py pyoqp/oqp/library/single_point.py`.
- `git diff --cached --check` passed before commit.

### CODE REVIEW

File: tools/scf-converger-ml/contribute/collect.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    ref = (g(r"\bscf\.?type\s*=?\s*(rhf|rohf|uhf)") or g(r"\b(RHF|ROHF|UHF) energy")).lower() or "rhf"
+    charge = g(r"charge\s*=?\s*(-?\d+)", "0")
+    conv = "convergence achieved" in t
+    es = re.findall(r"energy is\s+(-?\d+\.\d+)\s+after\s+(\d+)\s+iterations", t)
+    # TRAH macro 'micro' column -> add response builds to the cost
+    micro = sum(int(m.group(1)) for m in re.finditer(
+        r"^\s*\d+\s+-?\d+\.\d+\s+[\dEe.+-]+\s+[\dEe.+-]+\s+[\d.]+\s+(\d+)\s+\w+", t, re.M))
+    niter = (sum(int(n) for _, n in es) + micro) if es else -1
+    energy = es[-1][0] if es else "nan"
+    conver = g(r"Converger\s*=\s*([\w\-]+)") or g(r"diis type\s*:\s*(\w+)") or "cdiis"
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply tools/scf-converger-ml/contribute/collect.py:50]: Fixed in `17cad16`. The collector now strips the hyphen from native DIIS names (`conver.lower().replace("-", "")`), so `Converger = c-DIIS`/`e-DIIS`/`a-DIIS`/`v-DIIS` are written as `cdiis`/`ediis`/`adiis`/`vdiis` — the labels `_apply_selected_converger` actually maps.

File: tools/scf-converger-ml/contribute/collect.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  python contribute/collect.py --logs ~/my_openqp_runs --out data/contrib/yourname.csv --watch 60
+
+Then open a Pull Request adding your CSV (and any new geometries to geometries/). Hard cases
+where a converger FAILS are especially valuable. See README for the data policy.
+"""
+import argparse, csv, glob, os, re, time
+
+_TM = {'Sc','Ti','V','Cr','Mn','Fe','Co','Ni','Cu','Zn','Y','Zr','Nb','Mo','Tc','Ru','Rh','Pd','Ag','Cd',
+       'La','Hf','Ta','W','Re','Os','Ir','Pt','Au','Hg'}
+FIELDS = ["system","ref","func","converger","natom","charge","transition_metal","converged","niter","energy","basis"]
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply tools/scf-converger-ml/contribute/collect.py]: Fixed in `17cad16`. Added a `mult` column parsed from `PyOQP scf multiplicity` (regex `scf multiplicity\s*[:=]?\s*(\d+)`, default 1). Open-shell UHF/ROHF rows now carry their true multiplicity instead of being trained as singlets.

File: tools/scf-converger-ml/harness/build_selector.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        cells[(r["system"], r["ref"], r["func"])].append(r)
+
+    # best (oracle) converger per cell, plus cdiis baseline cost
+    X, y, oracle_cost, cdiis_cost, keys = [], [], [], [], []
+    wins = defaultdict(int)
+    for key, crows in cells.items():
+        ok = [r for r in crows if r["converged"] and r["niter"] > 0]
+        if not ok:
+            continue
+        best = min(ok, key=lambda r: r["niter"])
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply tools/scf-converger-ml/harness/build_selector.py]: Fixed in `17cad16`. `load()` now parses `fock_builds` (falling back to `niter` only for older DBs without the column), and the oracle/best, cdiis-baseline, and selector costs all use it. TRAH's expensive micro/response builds are no longer hidden behind its cheap macro count. Verified on `data/db_openqp_postfix.csv`: oracle=1014 vs always-cdiis=1524 Fock builds, TRAH wins only 1 cell.

File: tools/scf-converger-ml/contribute/collect.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+def parse_log(path):
+    t = open(path, errors="ignore").read()
+    if "MODULE" not in t and "energy is" not in t:
+        return None
+    def g(pat, d=""):
+        m = re.search(pat, t, re.I); return m.group(1).strip() if m else d
+    basis = g(r"basis\s*=?\s*([\w\-\*\(\)]+)")
+    func = g(r"functional\s*=?\s*([\w\-]+)") or "hf"
+    ref = (g(r"\bscf\.?type\s*=?\s*(rhf|rohf|uhf)") or g(r"\b(RHF|ROHF|UHF) energy")).lower() or "rhf"
+    charge = g(r"charge\s*=?\s*(-?\d+)", "0")
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply tools/scf-converger-ml/contribute/collect.py]: Fixed in `17cad16`. The metadata regexes now accept the colon form (`[:=]?`) and the `hf/functional` prefix, matching `dump_log`'s `PyOQP charge:`/`hf/functional:`/`basis:`/`scf type:` output. Verified: a DFT (`b3lypv5`), charge +3, UHF mult 6 log parses to `func='b3lypv5', charge='3', ref='uhf', mult='6'` instead of defaulting to hf/0.

File: tools/scf-converger-ml/harness/build_database.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    dt = time.time() - t
+    txt = open(log).read() if os.path.exists(log) else ""
+    es = re.findall(r"energy is\s+(-?\d+\.\d+)\s+after\s+(\d+)\s+iterations", txt)
+    conv = "convergence achieved" in txt
+    niter = sum(int(n) for _, n in es) if es else -1
+    E = es[-1][0] if es else "nan"
+    # True Fock-equivalent cost: each SCF iteration is ~1 Fock build; TRAH adds one
+    # response build per micro-iteration. DIIS/SOSCF: micro=0 so fock_builds==niter.
+    micro = trah_micro_sum(txt) if c == "trah" else 0
+    fock_builds = (niter + micro) if niter > 0 else -1
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply tools/scf-converger-ml/harness/build_database.py]: Fixed in `17cad16`. `micro = trah_micro_sum(txt)` is now called unconditionally (it returns 0 when the log has no TRAH table), so a DIIS/SOSCF primary that falls through to a TRAH escalation stage has that stage's micro builds counted toward `fock_builds`.

File: tools/scf-converger-ml/train/train_selector.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        that OpenQP's converger_type=ml loads at runtime.
+
+Usage : python train/train_selector.py --db data/db_openqp_postfix.csv
+"""
+import argparse, csv, collections, os, glob
+
+TM = set(range(21, 31)) | set(range(39, 49)) | set(range(57, 81))
+_SYM2Z = {  # minimal symbol->Z for TM detection from geometries
+    'Sc':21,'Ti':22,'V':23,'Cr':24,'Mn':25,'Fe':26,'Co':27,'Ni':28,'Cu':29,'Zn':30,
+    'Y':39,'Zr':40,'Nb':41,'Mo':42,'Tc':43,'Ru':44,'Rh':45,'Pd':46,'Ag':47,'Cd':48}
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply tools/scf-converger-ml/train/train_selector.py]: Fixed in `17cad16`. `_SYM2Z` now covers the same ranges as the runtime `_scf_features` TM set (21-30, 39-48, 57-80), i.e. through Hg. Verified `has_tm` returns 1 for Hf and Au systems; the map now has exactly 44 symbols matching the 44 runtime TM Z values.

File: tools/scf-converger-ml/harness/build_database.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            csv.DictWriter(f, fieldnames=fields).writeheader()
+
+    jobs = []
+    for s in SYSTEMS:
+        if s["tier"] > a.maxtier:
+            continue
+        natom = len(read_xyz(root, s["geom"]))
+        for ref, mult in s["refs"].items():
+            for func in FUNCS:
+                for c in CONVERGERS:
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply tools/scf-converger-ml/harness/build_database.py]: Fixed in `17cad16`. The functional and converger loops are now driven by `PROFILES[s['profile']]` instead of the two hard-coded globals (removed), so per-profile coverage such as `pbe`/`bhhlyp`/`vdiis` takes effect. `conv_block` gained a `vdiis` branch (`diis_type=vdiis`). Note: the guess dimension isn't a loop in this harness (inputs are fixed to huckel), so only functionals/convergers are profile-driven here.

File: tools/scf-converger-ml/contribute/collect.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                charge=charge, transition_metal=tm, converged=conv, niter=niter,
+                energy=energy, basis=basis)
+
+
+def collect(logs, out):
+    os.makedirs(os.path.dirname(out) or ".", exist_ok=True)
+    seen = set()
+    if os.path.exists(out):
+        for r in csv.DictReader(open(out)):
+            seen.add((r["system"], r["ref"], r["func"], r["converger"]))
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply tools/scf-converger-ml/contribute/collect.py]: Fixed in `17cad16`. The de-dup key is now `(system, ref, mult, func, converger, charge, basis)` — all run-defining fields — so the same molecule harvested at a different charge, multiplicity, or basis is kept as a distinct row. Verified keys differing only in charge/mult/basis are treated as distinct.
